File: frontend/src/main/java/al/recipes/controllers/HomeController.java
package al.recipes.controllers;

import al.recipes.models.Recipes;
import al.recipes.models.Tags;
import al.recipes.rest.controllers.TagControllerRest;
import al.recipes.soap.SoapClient;
import categories.wsdl.Categories;
import categories.wsdl.GetCategoriesResponse;
import io.swagger.annotations.ApiOperation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.MessageSource;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.hateoas.PagedResources;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.client.RestTemplate;

import java.util.*;

@Controller
@RequestMapping("/")
public class HomeController {
    
    private static final Logger log = LoggerFactory.getLogger(HomeController.class);
    @Autowired
    private SoapClient categorySoapClient;
    @Autowired
    private TagControllerRest tagControllerRest;
    @Autowired
    private MessageSource messageSource;
    
    @GetMapping(value = {"/", "/{page:[0-9]+}", "/{page:[0-9]+}/cat/{cat}", "/search/{keyword}"})
    @ApiOperation(value = "Get recipes by page / by page and by cat / search", notes = "Get recipes by page / by page and by cat / search")
    public String getAllRecipes(Model model, @PathVariable(value = "page") Optional<Integer> p, @PathVariable Optional<Integer> cat,
                                @PathVariable(value = "keyword") Optional<String> searchFilter, Locale locale) {
        
        final String header_txt = messageSource.getMessage("header_txt", null, locale);
        final String search_results = messageSource.getMessage("search_results", null, locale);
        
        List<Recipes> recipes = new ArrayList<>();
        List<String> tags = new ArrayList<>();
        List<Categories> categories = new ArrayList<>();
        List<Recipes> recent_recipes = new ArrayList<>();
        ResponseEntity<PagedResources<Recipes>> response = null;
        ResponseEntity<List<Recipes>> response_list = null;
        RestTemplate restTemplate = new RestTemplate();
        GetCategoriesResponse soapResponse = categorySoapClient.getCategories();
        categories.addAll(soapResponse.getCategories());
        Categories currentCat = null;
        Integer page = 1;
        if (p.isPresent()) page = p.get();
        String main_url = "https://bootiful-blog.herokuapp.com/api/recipes/" + page;
        String next_url = "https://bootiful-blog.herokuapp.com/api/recipes/" + (page + 1);
        String link_more = "/" + (page + 1);
        String tags_url = "https://bootiful-blog.herokuapp.com/api/tags";
        String recent_url = "https://bootiful-blog.herokuapp.com/api/recipes/1";
        
        if (cat.isPresent()) {
            main_url = "https://bootiful-blog.herokuapp.com/api/recipes/" + page + "/cat/" + cat.get();
            next_url = "https://bootiful-blog.herokuapp.com/api/recipes/" + (page + 1) + "/cat/" + cat.get();
            link_more = "/" + (page + 1) + "/cat/" + cat.get();
            currentCat = soapResponse.getCategories().stream().filter(c -> c.getId() == cat.get()).findFirst().get();
        }
        
        if (searchFilter.isPresent()) {
            main_url = "https://bootiful-blog.herokuapp.com/api/search/" + searchFilter.get();
            next_url = "";
            link_more = "#";
            currentCat = null;
            
            response_list =
                    restTemplate.exchange(main_url,
                            HttpMethod.GET, null, new ParameterizedTypeReference<List<Recipes>>() {
                            }
                    );
            Collection<Recipes> recipes_arr = Objects.requireNonNull(response_list.getBody());
            recipes_arr.stream().limit(20).forEach(r -> {
                r = setTagsIntro(r);
                recipes.add(r);
            });
            
        } else {
            
            response =
                    restTemplate.exchange(main_url,
                            HttpMethod.GET, null, new ParameterizedTypeReference<PagedResources<Recipes>>() {
                            }
                    );
            Collection<Recipes> recipes_arr = Objects.requireNonNull(response.getBody()).getContent();
            recipes_arr.stream().limit(20).forEach(r -> {
                r = setTagsIntro(r);
                recipes.add(r);
            });
            
        }
        
        
        ResponseEntity<PagedResources<Recipes>> recent_response =
                restTemplate.exchange(recent_url,
                        HttpMethod.GET, null, new ParameterizedTypeReference<PagedResources<Recipes>>() {
                        }
                );
        Collection<Recipes> recent_recipes_arr = Objects.requireNonNull(recent_response.getBody()).getContent();
        recent_recipes_arr.stream().limit(10).forEach(recent_recipes::add);
        
        ResponseEntity<List<String>> tags_response =
                restTemplate.exchange(tags_url,
                        HttpMethod.GET, null, new ParameterizedTypeReference<List<String>>() {
                        }
                );
        Collection<String> tags_arr = Objects.requireNonNull(tags_response.getBody());
        tags_arr.stream().limit(20).forEach(tags::add);
        
        if (!next_url.isEmpty()) {
            ResponseEntity<PagedResources<Recipes>> response_next =
                    restTemplate.exchange(next_url,
                            HttpMethod.GET, null, new ParameterizedTypeReference<PagedResources<Recipes>>() {
                            }
                    );
            
            if (Objects.requireNonNull(response_next.getBody()).getContent().size() > 0) {
                model.addAttribute("link_more", link_more);
            } else {
                model.addAttribute("link_more", "#");
            }
        } else {
            model.addAttribute("link_more", "#");
        }
        if (currentCat != null) {
            model.addAttribute("currentCat", currentCat);
        }
        
        searchFilter.ifPresent(s -> model.addAttribute("searchFilter", s));
        
        model.addAttribute("recent", recent_recipes);
        model.addAttribute("tagCloud", tags);
        model.addAttribute("recipes", recipes);
        model.addAttribute("categories", categories);
        if (searchFilter.isPresent()) {
            model.addAttribute("page_title", search_results + " " + searchFilter.get());
        } else if (currentCat != null) {
            model.addAttribute("page_title", currentCat.getName());
        } else {
            model.addAttribute("page_title", header_txt);
        }
        return "home";
    }
    
    private Recipes setTagsIntro(Recipes p) {
        RestTemplate restTemplate = new RestTemplate();
        Recipes recipe = restTemplate.getForObject("https://bootiful-blog.herokuapp.com/api/recipe/" + p.getId(), Recipes.class, 200);
        List<Tags> tags = Objects.requireNonNull(recipe).getTags();
        tags.sort(Comparator.comparingInt(Tags::getStart_pos));
        
        int start_pos = 0;
        int start_pos2 = 0;
        for (Tags tag : tags) {
            if (tag.getIntro_instruction().equals("intro")) {
                StringBuilder ins = new StringBuilder(p.getIntro());
                if (!ins.substring(tag.getStart_pos() + start_pos, tag.getEnd_pos() + start_pos).contains("<a") &&
                        !ins.substring(tag.getStart_pos() + start_pos2, tag.getEnd_pos() + start_pos2).contains("<a")) {
                    
                    p.setIntro(ins.replace(
                            (tag.getStart_pos() + start_pos),
                            (tag.getEnd_pos() + start_pos),
                            "<a class='tags' href='/search/" + tag.getValue() + "'>" + tag.getValue() + "</a>"
                    ).toString());
                    start_pos2 = start_pos;
                    start_pos += ("<a class='tags' href='/search/" + tag.getValue() + "'>" + "</a>").length();
                }
            }
        }
        return p;
    }

    /*private String translate(String txtToBeTranslated) {
        String txt = txtToBeTranslated;
        if (LocaleContextHolder.getLocale().toString().equals("en")) {
            RestTemplate restTemplate = new RestTemplate();
            ResponseEntity<String> response = restTemplate.getForEntity("https://www.googleapis.com/language/translate/" +
                    "v2?key=AIzaSyCbxc7thCFh8PQ9fmNJ_171bO3-jy_euqA&source=sq&target=en&q=" + txtToBeTranslated, String.class);

            txt = response.getBody().replace("{\n" +
                    "  \"data\": {\n" +
                    "    \"translations\": [\n" +
                    "      {\n" +
                    "        \"translatedText\": \"", "").replace("\"\n" +
                    "      }\n" +
                    "    ]\n" +
                    "  }\n" +
                    "}", "");
        }
        if (txt.length() > 100) return txt.substring(0, 100) + "...";
        return txt;
    }*/
}

File: frontend/src/main/java/al/recipes/controllers/HomeController.kt
package al.recipes.controllers

import al.recipes.models.Recipes
import al.recipes.soap.SoapClient
import categories.wsdl.Categories
import io.swagger.annotations.ApiOperation
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.MessageSource
import org.springframework.hateoas.PagedResources
import org.springframework.http.HttpMethod
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.client.RestTemplate
import org.springframework.web.client.exchange
import java.util.*
import javax.servlet.http.HttpServletRequest


@Controller
@RequestMapping("/")
class HomeController {
    @Autowired
    private val categorySoapClient: SoapClient? = null
    @Autowired
    private val messageSource: MessageSource? = null

    @GetMapping(value = ["/", "/{page:[0-9]+}", "/{page:[0-9]+}/cat/{cat}", "/search/{keyword}"])
    @ApiOperation(value = "Get recipes by page / by page and by cat / search", notes = "Get recipes by page / by page and by cat / search")
    fun getAllRecipes(model: Model, @PathVariable(value = "page") pageRequest: Optional<Int>, @PathVariable cat: Optional<Int>,
                      @PathVariable(value = "keyword") searchFilter: Optional<String>, locale: Locale, request: HttpServletRequest): String {
        val headerTxt = messageSource!!.getMessage("header_txt", null, locale)
        val searchResults = messageSource.getMessage("search_results", null, locale)
        val https = request.scheme as String
        val urlConn = https + "://" + request.serverName
        val recipes = ArrayList<Recipes>()
        val tags = ArrayList<String>()
        val recentRecipes = ArrayList<Recipes>()
        val response: ResponseEntity<PagedResources<Recipes>>
        val responseList: ResponseEntity<List<Recipes>>
        val restTemplate = RestTemplate()
        var page = 1
        if (pageRequest.isPresent) {
            page = pageRequest.get()
        }
        var currentCat: Categories? = null
        var mainUrl = "$urlConn/api/recipes/$page"
        var nextUrl = urlConn + "/api/recipes/" + (page + 1)
        var moreLink = "/" + (page + 1)
        val tagsUrl = "$urlConn/api/tags"
        val recentUrl = "$urlConn/api/recipes/1"

        // Get list of categories (sidebar)
        val categories = ArrayList<Categories>()
        val soapResponse = categorySoapClient!!.categories
        categories.addAll(soapResponse.categories)

        // If showing recipes of a specific category, get category details
        if (cat.isPresent) {
            mainUrl = urlConn + "/api/recipes/" + page + "/cat/" + cat.get()
            nextUrl = urlConn + "/api/recipes/" + (page + 1) + "/cat/" + cat.get()
            moreLink = "/" + (page + 1) + "/cat/" + cat.get()
            currentCat = soapResponse.categories.stream().filter { c -> c.id == cat.get() }.findFirst().get()
        }

        // If search filter is present, filter recipes by searchFilter
        if (searchFilter.isPresent) {
            mainUrl = urlConn + "/api/search/" + searchFilter.get()
            nextUrl = ""
            moreLink = "#"
            currentCat = null

            responseList = restTemplate.exchange(mainUrl, HttpMethod.GET)
            val recipes_arr = responseList.getBody()
            recipes_arr?.stream()?.limit(20)?.forEach { recipes.add(setTagsIntro(it)) }
        } else {
            response = restTemplate.exchange(mainUrl, HttpMethod.GET)
            val recipes_arr = Objects.requireNonNull<PagedResources<Recipes>>(response.body).content
            recipes_arr.stream().limit(20).forEach { recipes.add(setTagsIntro(it)) }
        }

        // Get recent recipes (sidebar)
        val recentRecipesResponse: ResponseEntity<PagedResources<Recipes>> = restTemplate.exchange(recentUrl, HttpMethod.GET)
        val recentRecipesArray = Objects.requireNonNull<PagedResources<Recipes>>(recentRecipesResponse.body).content
        recentRecipesArray.stream().limit(10).forEach { recentRecipes.add(it) }

        // Get tag cloud (sidebar)
        val tagCloudResponse: ResponseEntity<List<String>> = restTemplate.exchange(tagsUrl, HttpMethod.GET)
        val tagCloudArray = Objects.requireNonNull<List<String>>(tagCloudResponse.body)
        tagCloudArray.stream().limit(20).forEach { tags.add(it) }

        if (!nextUrl.isEmpty()) {
            val response_next: ResponseEntity<PagedResources<Recipes>> = restTemplate.exchange(nextUrl, HttpMethod.GET)
            if (Objects.requireNonNull<PagedResources<Recipes>>(response_next.body).content.isNotEmpty()) {
                model.addAttribute("link_more", moreLink)
            } else {
                model.addAttribute("link_more", "#")
            }
        } else {
            model.addAttribute("link_more", "#")
        }
        if (currentCat != null) {
            model.addAttribute("currentCat", currentCat)
        }

        searchFilter.ifPresent { s ->
            model.addAttribute("searchFilter", s)
        }

        model.addAttribute("recent", recentRecipes)
        model.addAttribute("tagCloud", tags)
        model.addAttribute("recipes", recipes)
        model.addAttribute("categories", categories)
        when {
            searchFilter.isPresent -> model.addAttribute("page_title", searchResults + " " + searchFilter.get())
            currentCat != null -> model.addAttribute("page_title", currentCat.name)
            else -> model.addAttribute("page_title", headerTxt)
        }
        return "home"
    }


    private fun setTagsIntro(p: Recipes): Recipes {
        val tags = Objects.requireNonNull<Recipes>(p).tags
        tags.sortWith(Comparator.comparingInt { it.start_pos })

        var startPos = 0
        var startPos2 = 0
        for (tag in tags) {
            if (tag.intro_instruction == "intro") {
                val ins = StringBuilder(p.intro)
                if (!ins.substring(tag.start_pos + startPos, tag.end_pos + startPos).contains("<a") && !ins.substring(tag.start_pos + startPos2, tag.end_pos + startPos2).contains("<a")) {

                    p.intro = ins.replace(
                            tag.start_pos + startPos,
                            tag.end_pos + startPos,
                            "<a class='tags' href='/search/" + tag.value + "'>" + tag.value + "</a>"
                    ).toString()
                    startPos2 = startPos
                    startPos += ("<a class='tags' href='/search/" + tag.value + "'>" + "</a>").length
                }
            }
        }
        return p
    }
}

